### PR TITLE
Consolidate enum for observer step

### PR DESCRIPTION
**Pull request by namgyu-youn**
**Summary**:
Calibration-based algorithms (AWQ, SmoothQuant) operate independently of quantization granularity. To improve extensibility and consistency, add common enum instead of algorithm-specific definitions.

**Test plan**:
```bash
pytest -sv test/core/test_config.py
pytest -sv test/prototype/test_awq.py
pytest -sv test/prototype/test_smoothquant.py
```

**Example**: SmoothQuant ([arxiv:2211.10438](https://arxiv.org/abs/2211.10438)) decomposes into:

Quantization Flow
1. **Prepare**: Insert observers into linear layers
2. **Calibrate**: Compute scaling factors via calibration algorithm
3. **Convert**: Apply calibration results and quantize weights

Inference Flow: Load quantized model directly

### CODE REVIEW

File: torchao/quantization/observer.py [hunk 1]
```diff
@@ -346,3 +347,31 @@ def calculate_qparams(self):
...
+# after python 3.10 is end of life (https://devguide.python.org/versions/)
+class ObserverStep(str, Enum):
+    """
+    The step enum backend for observer based algorithms process like AWQ/SmoothQuant/GPTQ
+        PREPARE: insert observers to linear layers
+        CONVERT: convert the observed linear modules to quantized modules
+        PREPARE_FOR_LOADING: convert the floating point model to a dummy quantized model,
+        so we can load the quantized weights through copy_ later
+
+    Example:
```
Comment (jcaip) [inline torchao/quantization/observer.py]: nit: can you also add an example of `PREPARE_FOR_LOADING`?

**Comment by namgyu-youn** [comment]
> would you mind updating gptq.py also once #3517 lands?

Prefer to land this first for less LOC, but will open small PR if you want to land GPTQ first :)

**Comment by jcaip** [comment]
> > would you mind updating gptq.py also once #3517 lands?
> 
> Prefer to land this first for less LOC, but will open small PR if you want to land GPTQ first :)

Oh I didn't mean for you to wait, go ahead and land this first :)

File: test/core/test_config.py [hunk 2]
```diff
@@ -22,13 +22,13 @@
...
 from torchao.prototype.awq import (
     AWQConfig,
-    AWQStep,
 )
 from torchao.quantization import (
     PerBlock,
     PerRow,
     PerTensor,
 )
+from torchao.quantization.observer import ObserverStep
```
Comment (jerryzh168) [inline test/core/test_config.py]: I think it should not be ObserverStep, maybe QuantizationStep would be more appropriate I feel

File: torchao/quantization/observer.py [hunk 3]
```diff
@@ -346,3 +347,41 @@ def calculate_qparams(self):
             self.preserve_zero,
             self.zero_point_domain,
         )
+
+
+# can switch to StrEnum (https://docs.python.org/3/library/enum.html#enum.StrEnum)
+# after python 3.10 is end of life (https://devguide.python.org/versions/)
+class ObserverStep(str, Enum):
```
Comment (jerryzh168) [inline torchao/quantization/observer.py]: can you move this to https://github.com/pytorch/ao/tree/main/torchao/quantization/quantize_/common as well, after renaming, it's not really tied to the observer I think

Comment (namgyu-youn) [reply torchao/quantization/observer.py]: Moved to `torchao/quantization/quantize_/observer/utils.py`, because I thought following file structure would be easier to extend for future promotion (AWQ/SmoothQuant/GPTQ). Let me know if I should move to common folder or something instead.
```
├── __init__.py
├── common
├── observer
│   ├── __init__.py
│   ├── awq
│   ├── smoothquant
│   ├── gptq
│   └── utils.py
└── workflows - Int4Tensor, Int8Tensor, etc
```

Comment (jerryzh168) [reply torchao/quantization/observer.py]: some observers are actually shared in multiple flows (quantize_ and pt2e flow), that's why they are kept in torchao.quantization

for awq/smoothquant/gptq, I don't think we need a separate folder for observer, it's better to keep the observer definition in the same folder of these techniques itself, so it's easy to find all the implementations:

```
├── __init__.py
├── common
├── observer
│   ├── __init__.py
│   ├── awq
│   ├── smoothquant
│   ├── gptq
│   └── utils.py
└── workflows 
   * int4
   * int8
   * fp8
   * awq
     * observer
     * api
   * gptq
     * observer
     * api
   * smoothquant
     * observer
     * config
```

File: torchao/quantization/quantize_/common/quantization_step.py
Comment (jerryzh168) [inline torchao/quantization/quantize_/common/quantization_step.py:1]: it should be `torchao/quantization/quantize_/common/quantization_step.py` I think

File: torchao/quantization/observer.py (hunk 3 above)
Comment (namgyu-youn) [reply torchao/quantization/observer.py]: Okay, I think awq/smoothquant/gptq can be separated from workflows because those require calibration flow and observer, unlike upstream workflows (focusing on memory layout without calibration), but I will follow your decision.

File: torchao/quantization/quantize_/common/quantization_step.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        quantize_(model, Float8StaticActivationFloat8WeightConfig(step="prepare"))
+
+        # Stage 2: CALIBRATE - collect statistics
+        for batch in calibration_data:
+            model(batch)
+
+        # Stage 3: CONVERT - apply quantization and remove observers
+        quantize_(model, Float8StaticActivationFloat8WeightConfig(step="convert"))
+
+        # Stage 4 (optional): PREPARE_FOR_LOADING - for saving/loading quantized models
```
Comment (jerryzh168) [inline torchao/quantization/quantize_/common/quantization_step.py]: nit: this is not Stage 4, it's a separate flow actually, maybe you can make it clear by saying Flow 1 - Stage 1, Flow 1 - Stage 2, Flow 1 - Stage 3 and Flow2

Comment (namgyu-youn) [reply torchao/quantization/quantize_/common/quantization_step.py]: Precisely true, but isn't it too much info? 1-3 is the main workflow, so we can just label the save/load flow as optional I feel.

Comment (jerryzh168) [reply torchao/quantization/quantize_/common/quantization_step.py]: I think we should make it clearer, current way of structuring it is a bit confusing / incorrect

main thing is you don't need to run step 1-3, to run "prepare_for_loading" step, it's an independent step that can be run during loading.

by optional step 4, what is conveyed is that:

flow 1:
step 1, 2, 3

flow 2:
step 1, 2, 3, 4

but that is not correct

Comment (namgyu-youn) [reply torchao/quantization/quantize_/common/quantization_step.py]: Oh sorry, those should be separated to: Flow 1 (quantization; step 1-3) & Flow 2 (inference; step 4), right? Updated example to separate them, thanks for pointing it out.

**Comment by namgyu-youn** [comment]
Looks like there are some unrelated ruff issue: https://github.com/pytorch/ao/actions/runs/21311160018/job/61548361113?pr=3678

**Comment by jerryzh168** [comment]
@namgyu-youn can you rebase on main? I remember this is resolved already

**Comment by jerryzh168** [comment]
can you check docs as well? https://github.com/pytorch/ao/tree/main/docs#using-workflow_dispatch-for-doc-previews

File: torchao/quantization/quantize_/common/quantization_step.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        quantize_(model, config)
+
+        # Now the model can load pre-quantized weights
+        model.load_state_dict(quantized_state_dict)
+
+    """
+
+    PREPARE = "prepare"
+    CONVERT = "convert"
+    PREPARE_FOR_LOADING = "prepare_for_loading"
```
Comment (jerryzh168) [inline torchao/quantization/quantize_/common/quantization_step.py:46]: I feel it might be better to document each of these separately, can move the docs from L14 here

Comment (jerryzh168) [reply torchao/quantization/quantize_/common/quantization_step.py:46]: example https://docs.pytorch.org/ao/main/_modules/torchao/quantization/quantize_/common/kernel_preference.html#KernelPreference

Comment (namgyu-youn) [reply torchao/quantization/quantize_/common/quantization_step.py:46]: Okay, update via https://github.com/pytorch/ao/pull/3678/changes/e88b377a098c63465681e66460289822797e93a6

**Comment by namgyu-youn** [comment]
> can you check docs as well? [`main`/docs#using-workflow_dispatch-for-doc-previews](https://github.com/pytorch/ao/tree/main/docs?rgh-link-date=2026-01-27T23%3A11%3A39Z#using-workflow_dispatch-for-doc-previews)

Sure, but could you please look into other PRs first? There are https://github.com/pytorch/ao/pull/3148/changes, https://github.com/pytorch/ao/pull/3617/changes, https://github.com/pytorch/ao/pull/3687/changes which we have talked about before.

**Comment by jerryzh168** [comment]
> > can you check docs as well? [`main`/docs#using-workflow_dispatch-for-doc-previews](https://github.com/pytorch/ao/tree/main/docs?rgh-link-date=2026-01-27T23%3A11%3A39Z#using-workflow_dispatch-for-doc-previews)
> 
> Sure, but could you please look into other PRs first? There are [#3148 (changes)](https://github.com/pytorch/ao/pull/3148/changes), [#3617 (changes)](https://github.com/pytorch/ao/pull/3617/changes), [#3687 (changes)](https://github.com/pytorch/ao/pull/3687/changes) which we have talked about before.

sure, I was talking about checking the doc built for this PR in case there is some misunderstanding

**Comment by jerryzh168** [comment]
oh actually the doc action only works when the commit is in the pytorch repo, we should probably fix it, I'll merge this PR for now

File: torchao/quantization/quantize_/common/quantization_step.py (hunk 5 above)
Comment (jerryzh168) [reply torchao/quantization/quantize_/common/quantization_step.py:46]: edit: please ignore previous comment for now, I actually haven't find docs for how to people really document enum with sphinx yet
